File: sdks/dotnet/ReproIt.Sdk/ProcessorCapture.cs
// SDK-side processor capture.
//
// The canonical Linux capture rule is pinned by
// specs/v1/processor-capture.json. Every SDK derives the same sorted
// capability list from the same host. Capture maps raw views through
// curated tables and never removes a captured value. A non-Linux host
// captures nothing. A read or parse failure captures nothing, because
// capture may only add real information and a failed SDK must never change
// application behavior.

using System;
using System.Buffers.Binary;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;

namespace ReproIt.Sdk;

/// <summary>Captures the process-visible processor view of this host.</summary>
public static class ProcessorCapture
{
    private const string FeaturePrefix = "processor.feature.";
    private const string OsStatePrefix = "processor.os-state.";
    private const string IdentityPrefix = "processor.identity.";
    private const ulong AtHwcap = 16;

    private static readonly Dictionary<string, string> X86FlagGroups = new()
    {
        ["aes"] = "aes",
        ["avx"] = "avx",
        ["avx2"] = "avx2",
        ["avx512bw"] = "avx512bw",
        ["avx512dq"] = "avx512dq",
        ["avx512f"] = "avx512f",
        ["avx512vl"] = "avx512vl",
        ["bmi1"] = "bmi1",
        ["bmi2"] = "bmi2",
        ["fma"] = "fma",
        ["pclmulqdq"] = "pclmulqdq",
        ["pni"] = "sse3",
        ["sse2"] = "sse2",
        ["sse4_1"] = "sse4-1",
        ["sse4_2"] = "sse4-2",
        ["ssse3"] = "ssse3",
    };

    private static readonly Dictionary<int, string> Arm64HwcapGroups = new()
    {
        [1] = "asimd",
        [3] = "aes",
        [6] = "sha2",
        [7] = "crc32",
        [22] = "sve",
    };

    private static readonly string[] X86IdentityFields =
        ["vendor_id", "cpu family", "model", "stepping", "microcode"];

    private static readonly string[] ArmIdentityFields =
        ["CPU implementer", "CPU variant", "CPU part", "CPU revision"];

    /// <summary>Captures sorted processor.* capability strings.</summary>
    public static List<string> CaptureProcessorCapabilities()
    {
        if (!OperatingSystem.IsLinux())
        {
            return [];
        }
        string machine = RuntimeInformation.ProcessArchitecture switch
        {
            Architecture.X64 => "x86_64",
            Architecture.Arm64 => "aarch64",
            _ => "",
        };
        if (machine.Length == 0)
        {
            return [];
        }
        string cpuinfo;
        try
        {
            cpuinfo = File.ReadAllText("/proc/cpuinfo");
        }
        catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
        {
            cpuinfo = "";
        }
        ulong? hwcap = null;
        try
        {
            hwcap = ParseAuxvHwcap(File.ReadAllBytes("/proc/self/auxv"));
        }
        catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
        {
            hwcap = null;
        }
        return DeriveProcessorCapabilities(machine, cpuinfo, hwcap);
    }

    /// <summary>Pure derivation over raw views, shared with the capture
    /// vectors.</summary>
    public static List<string> DeriveProcessorCapabilities(
        string machine, string cpuinfo, ulong? hwcap)
    {
        SortedSet<string> unique = new(StringComparer.Ordinal);
        string? identity;
        if (machine == "x86_64")
        {
            HashSet<string> flags = X86Flags(cpuinfo);
            foreach ((string flag, string group) in X86FlagGroups)
            {
                if (flags.Contains(flag))
                {
                    unique.Add(FeaturePrefix + group);
                }
            }
            if (flags.Contains("osxsave"))
            {
                unique.Add(OsStatePrefix + "osxsave");
            }
            if (flags.Contains("avx"))
            {
                unique.Add(OsStatePrefix + "xcr0.avx");
            }
            if (flags.Contains("avx512f"))
            {
                unique.Add(OsStatePrefix + "xcr0.avx512");
            }
            identity = IdentityToken(cpuinfo, X86IdentityFields, "x86", [1, 2, 3]);
        }
        else if (machine == "aarch64")
        {
            if (hwcap is ulong value)
            {
                foreach ((int bit, string group) in Arm64HwcapGroups)
                {
                    if ((value & (1UL << bit)) != 0)
                    {
                        unique.Add(FeaturePrefix + group);
                    }
                }
                unique.Add(OsStatePrefix + "auxv.hwcaps");
            }
            identity = IdentityToken(cpuinfo, ArmIdentityFields, "arm", []);
        }
        else
        {
            return [];
        }
        if (identity is not null)
        {
            unique.Add(IdentityPrefix + identity);
        }
        return [.. unique];
    }

    /// <summary>Extracts AT_HWCAP from raw /proc/self/auxv bytes:
    /// little-endian 16-byte entries of key then value, terminated by a
    /// zero key.</summary>
    public static ulong? ParseAuxvHwcap(byte[] auxv)
    {
        for (int offset = 0; offset + 16 <= auxv.Length; offset += 16)
        {
            ulong key = BinaryPrimitives.ReadUInt64LittleEndian(auxv.AsSpan(offset, 8));
            if (key == 0)
            {
                return null;
            }
            if (key == AtHwcap)
            {
                return BinaryPrimitives.ReadUInt64LittleEndian(auxv.AsSpan(offset + 8, 8));
            }
        }
        return null;
    }

    private static HashSet<string> X86Flags(string cpuinfo)
    {
        HashSet<string> flags = new(StringComparer.Ordinal);
        foreach (string line in cpuinfo.Split('\n'))
        {
            int separator = line.IndexOf(':', StringComparison.Ordinal);
            if (separator < 0 || line[..separator].Trim() != "flags")
            {
                continue;
            }
            foreach (string flag in line[(separator + 1)..]
                .Split(' ', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries))
            {
                flags.Add(flag);
            }
        }
        return flags;
    }

    /// <summary>Encodes the canonical lowercase identity token, or null
    /// when a field is missing, malformed, or disagrees between processor
    /// blocks.</summary>
    private static string? IdentityToken(
        string cpuinfo, string[] fields, string prefix, int[] numeric)
    {
        List<List<string>> blocks = [];
        foreach (string block in cpuinfo.Split("\n\n"))
        {
            if (block.Trim().Length == 0)
            {
                continue;
            }
            List<string> values = [];
            foreach (string field in fields)
            {
                string? found = null;
                foreach (string line in block.Split('\n'))
                {
                    int separator = line.IndexOf(':', StringComparison.Ordinal);
                    if (separator >= 0 && line[..separator].Trim() == field)
                    {
                        found = line[(separator + 1)..].Trim();
                        break;
                    }
                }
                if (string.IsNullOrEmpty(found))
                {
                    return null;
                }
                values.Add(found);
            }
            blocks.Add(values);
        }
        if (blocks.Count == 0 || blocks.Any(block => !block.SequenceEqual(blocks[0])))
        {
            return null;
        }
        List<string> lowered = [.. blocks[0].Select(value => value.ToLowerInvariant())];
        if (lowered.Any(value =>
            value.Any(character => character is not ((>= 'a' and <= 'z')
                or (>= '0' and <= '9') or '-'))))
        {
            return null;
        }
        foreach (int index in numeric)
        {
            if (!uint.TryParse(lowered[index], NumberStyles.None, CultureInfo.InvariantCulture,
                out _))
            {
                return null;
            }
        }
        return string.Join('.', lowered.Prepend(prefix));
    }
}

File: sdks/go/reproit/processor_capture.go
package reproit

// SDK-side processor capture.
//
// The canonical Linux capture rule is pinned by
// specs/v1/processor-capture.json. Every SDK derives the same sorted
// capability list from the same host. Capture maps raw views through
// curated tables and never removes a captured value. A non-Linux host
// captures nothing. A read or parse failure captures nothing, because
// capture may only add real information and a failed SDK must never change
// application behavior.

import (
	"encoding/binary"
	"os"
	"runtime"
	"sort"
	"strconv"
	"strings"
)

const (
	processorFeaturePrefix  = "processor.feature."
	processorOsStatePrefix  = "processor.os-state."
	processorIdentityPrefix = "processor.identity."
	atHwcap                 = 16
)

var x86FlagGroups = map[string]string{
	"aes":       "aes",
	"avx":       "avx",
	"avx2":      "avx2",
	"avx512bw":  "avx512bw",
	"avx512dq":  "avx512dq",
	"avx512f":   "avx512f",
	"avx512vl":  "avx512vl",
	"bmi1":      "bmi1",
	"bmi2":      "bmi2",
	"fma":       "fma",
	"pclmulqdq": "pclmulqdq",
	"pni":       "sse3",
	"sse2":      "sse2",
	"sse4_1":    "sse4-1",
	"sse4_2":    "sse4-2",
	"ssse3":     "ssse3",
}

var arm64HwcapGroups = map[uint]string{
	1:  "asimd",
	3:  "aes",
	6:  "sha2",
	7:  "crc32",
	22: "sve",
}

var x86IdentityFields = []string{"vendor_id", "cpu family", "model", "stepping", "microcode"}

var armIdentityFields = []string{"CPU implementer", "CPU variant", "CPU part", "CPU revision"}

// CaptureProcessorCapabilities captures this host's process-visible
// processor view as sorted processor.* capability strings.
func CaptureProcessorCapabilities() []string {
	if runtime.GOOS != "linux" {
		return nil
	}
	machine := ""
	switch runtime.GOARCH {
	case "amd64":
		machine = "x86_64"
	case "arm64":
		machine = "aarch64"
	default:
		return nil
	}
	cpuinfo := ""
	if raw, err := os.ReadFile("/proc/cpuinfo"); err == nil {
		cpuinfo = string(raw)
	}
	var hwcap *uint64
	if raw, err := os.ReadFile("/proc/self/auxv"); err == nil {
		hwcap = parseAuxvHwcap(raw)
	}
	return deriveProcessorCapabilities(machine, cpuinfo, hwcap)
}

// deriveProcessorCapabilities is the pure derivation over raw views, shared
// with the capture vectors.
func deriveProcessorCapabilities(machine, cpuinfo string, hwcap *uint64) []string {
	unique := map[string]bool{}
	var identity string
	switch machine {
	case "x86_64":
		flags := x86Flags(cpuinfo)
		for flag, group := range x86FlagGroups {
			if flags[flag] {
				unique[processorFeaturePrefix+group] = true
			}
		}
		if flags["osxsave"] {
			unique[processorOsStatePrefix+"osxsave"] = true
		}
		if flags["avx"] {
			unique[processorOsStatePrefix+"xcr0.avx"] = true
		}
		if flags["avx512f"] {
			unique[processorOsStatePrefix+"xcr0.avx512"] = true
		}
		identity = identityToken(cpuinfo, x86IdentityFields, "x86", []int{1, 2, 3})
	case "aarch64":
		if hwcap != nil {
			for bit, group := range arm64HwcapGroups {
				if *hwcap&(uint64(1)<<bit) != 0 {
					unique[processorFeaturePrefix+group] = true
				}
			}
			unique[processorOsStatePrefix+"auxv.hwcaps"] = true
		}
		identity = identityToken(cpuinfo, armIdentityFields, "arm", nil)
	default:
		return nil
	}
	if identity != "" {
		unique[processorIdentityPrefix+identity] = true
	}
	capabilities := make([]string, 0, len(unique))
	for capability := range unique {
		capabilities = append(capabilities, capability)
	}
	sort.Strings(capabilities)
	return capabilities
}

// parseAuxvHwcap extracts AT_HWCAP from raw /proc/self/auxv bytes:
// little-endian 16-byte entries of key then value, terminated by a zero key.
func parseAuxvHwcap(auxv []byte) *uint64 {
	for offset := 0; offset+16 <= len(auxv); offset += 16 {
		key := binary.LittleEndian.Uint64(auxv[offset:])
		if key == 0 {
			return nil
		}
		if key == atHwcap {
			value := binary.LittleEndian.Uint64(auxv[offset+8:])
			return &value
		}
	}
	return nil
}

func x86Flags(cpuinfo string) map[string]bool {
	flags := map[string]bool{}
	for _, line := range strings.Split(cpuinfo, "\n") {
		name, value, found := strings.Cut(line, ":")
		if !found || strings.TrimSpace(name) != "flags" {
			continue
		}
		for _, flag := range strings.Fields(value) {
			flags[flag] = true
		}
	}
	return flags
}

// identityToken encodes the canonical lowercase identity token, or empty
// when a field is missing, malformed, or disagrees between processor
// blocks.
func identityToken(cpuinfo string, fields []string, prefix string, numeric []int) string {
	var blocks [][]string
	for _, block := range strings.Split(cpuinfo, "\n\n") {
		if strings.TrimSpace(block) == "" {
			continue
		}
		values := make([]string, 0, len(fields))
		for _, field := range fields {
			found := ""
			for _, line := range strings.Split(block, "\n") {
				name, value, ok := strings.Cut(line, ":")
				if ok && strings.TrimSpace(name) == field {
					found = strings.TrimSpace(value)
					break
				}
			}
			if found == "" {
				return ""
			}
			values = append(values, found)
		}
		blocks = append(blocks, values)
	}
	if len(blocks) == 0 {
		return ""
	}
	for _, block := range blocks[1:] {
		for index, value := range block {
			if value != blocks[0][index] {
				return ""
			}
		}
	}
	lowered := make([]string, len(blocks[0]))
	for index, value := range blocks[0] {
		lowered[index] = strings.ToLower(value)
		for _, character := range lowered[index] {
			if (character < 'a' || character > 'z') &&
				(character < '0' || character > '9') && character != '-' {
				return ""
			}
		}
	}
	for _, index := range numeric {
		if _, err := strconv.ParseUint(lowered[index], 10, 32); err != nil {
			return ""
		}
	}
	return prefix + "." + strings.Join(lowered, ".")
}
